src/libPMacc/include/math/MapTuple.hpp

```diff
@@ -23,199 +23,215 @@
-    at()
-    {
-        return (*this)[
-            typename mpl::deref<
-            typename mpl::advance<
-            typename mpl::begin<Map>::type, mpl::int_<i> >::type>::type::first()];
-    }
-};
-
-#undef CONSTRUCTOR
-
-#define PAIR_LIST(Z, N, _) mpl::pair<Key ## N, Value ## N>
-
-#define MAKE_MAPTUPLE(Z, N, _) \
-    template<BOOST_PP_ENUM_PARAMS(N, typename Key), BOOST_PP_ENUM_PARAMS(N, typename Value)> \
-    HDINLINE \
-    MapTuple<mpl::map<BOOST_PP_ENUM(N, PAIR_LIST, _)> > \
-    make_MapTuple(BOOST_PP_ENUM_BINARY_PARAMS(N, const Value, &value)) \
-    { \
-        return MapTuple<mpl::map<BOOST_PP_ENUM(N, PAIR_LIST, _)> > \
-            (BOOST_PP_ENUM_PARAMS(N, value)); \
-    }
-
-BOOST_PP_REPEAT_FROM_TO(1, BOOST_PP_INC(MAPTUPLE_MAX_DIM), MAKE_MAPTUPLE, _)
-
-#undef MAKE_MAPTUPLE
-#undef PAIR_LIST
+        typedef Map_ Map;
+        BOOST_STATIC_CONSTEXPR int dim = bmpl::size< Map >::type::value;
+        typedef InheritLinearly<
+            Map_,
+            PODType
+        > Base;
+
+        template<class> struct result;
+
+        template<class F, class TKey>
+        struct result<F(TKey)>
+        {
+            typedef typename bmpl::at<Map, TKey>::type& type;
+        };
+
+        template<class F, class TKey>
+        struct result<const F(TKey)>
+        {
+            typedef const typename bmpl::at<Map, TKey>::type& type;
+        };
+
+        /** access a datum with a key
+         *
+         * @tparam T_Key key type
+         *
+         * @{
+         */
+        template< typename T_Key >
+        HDINLINE
+        typename boost::result_of<
+            MapTuple(T_Key)
+        >::type
+        operator[](const T_Key& key)
+        {
+            return
+            (
+                *(static_cast<
+                    PODType<
+                        bmpl::pair<
+                            T_Key,
+                            typename bmpl::at<
+                                Map,
+                                T_Key
+                            >::type
+                        >
+                    >*
+                >(this))
+            )[key];
+        }
+
+        template< typename T_Key >
+        HDINLINE
+        typename boost::result_of<
+            const MapTuple(T_Key)
+        >::type
+        operator[](const T_Key& key) const
+        {
+            return (
+                *(
+                    static_cast<
+                        const PODType<
+                            bmpl::pair<
+                                T_Key,
+                                typename bmpl::at<
+                                    Map,
+                                    T_Key
+                                >::type
+                            >
+                        >*
+                    >(this)
+                )
+            )[key];
+        }
+        /** @} */
+
+        /** access a datum with an index
+         *
+         * @tparam i index
```

Comment:
you can be more precise about `T_i` here, something like `@tparam T_i the index of tuple's i-th element`